File: src/agent_nexus/platform/agency/cli_backend/parser.py
"""CLI output parser — JSON path extraction and text regex parsing."""

from __future__ import annotations

import json
import re
from typing import Any

from agent_nexus.platform.agency.cli_backend.types import BackendConfig, CLIResult


def extract_json_value(data: dict[str, Any], path: str) -> Any:
    """Extract a value from a nested dict using dot-separated path."""
    if not path:
        return None
    keys = path.split(".")
    current: Any = data
    for key in keys:
        if not isinstance(current, dict) or key not in current:
            return None
        current = current[key]
    return current


def parse_json_output(stdout: str, config: BackendConfig) -> CLIResult:
    """Parse JSON-formatted CLI output using json_paths config."""
    try:
        data = json.loads(stdout)
    except (json.JSONDecodeError, TypeError):
        return CLIResult(text=stdout, model="", raw_stdout=stdout, parse_error=True)

    paths = config.json_paths

    def _extract(path: str | None) -> Any:
        if path is None:
            return None
        return extract_json_value(data, path)

    text = _extract(paths.text)
    if text is None:
        text = stdout
    if not isinstance(text, str):
        text = str(text)

    session_id = _extract(paths.session_id)
    model = _extract(paths.model)
    input_tokens = _extract(paths.input_tokens)
    output_tokens = _extract(paths.output_tokens)

    return CLIResult(
        text=text,
        model=model if isinstance(model, str) else "",
        session_id=session_id if isinstance(session_id, str) else None,
        input_tokens=int(input_tokens) if isinstance(input_tokens, (int, float)) else None,
        output_tokens=int(output_tokens) if isinstance(output_tokens, (int, float)) else None,
        raw_stdout=stdout,
        parse_error=False,
    )


def parse_text_output(stdout: str, stderr: str, config: BackendConfig) -> CLIResult:
    """Parse text-mode CLI output, optionally extracting metadata via regex."""
    session_id = None
    model = None
    patterns = config.text_patterns
    if patterns.session_id:
        combined = f"{stdout}\n{stderr}"
        match = re.search(patterns.session_id, combined)
        if match:
            session_id = match.group(1)
    if patterns.model:
        match = re.search(patterns.model, stdout)
        if match:
            model = match.group(1)
    return CLIResult(
        text=stdout,
        model=model or "",
        session_id=session_id,
        raw_stdout=stdout,
        raw_stderr=stderr,
        parse_error=False,
    )

File: crates/ap-cli-backend/src/parser.rs
//! Output parsing — JSON dot-path extraction and text regex extraction.

use crate::types::{BackendConfig, CLIResult, JsonPathConfig};

pub fn extract_json_value<'a>(data: &'a serde_json::Value, path: &str) -> Option<&'a serde_json::Value> {
    let mut current = data;
    for key in path.split('.') {
        match current {
            serde_json::Value::Object(map) => {
                current = map.get(key)?;
            }
            _ => return None,
        }
    }
    Some(current)
}

pub fn parse_json_output(stdout: &str, config: &BackendConfig) -> CLIResult {
    let mut result = CLIResult {
        raw_stdout: stdout.to_string(),
        ..Default::default()
    };

    let json_data: serde_json::Value = match serde_json::from_str(stdout.trim()) {
        Ok(v) => v,
        Err(_) => {
            result.text = stdout.to_string();
            result.parse_error = true;
            return result;
        }
    };

    build_result_from_json(&json_data, &config.json_paths, &mut result);
    result.raw_stdout = stdout.to_string();
    result
}

fn build_result_from_json(
    data: &serde_json::Value,
    paths: &JsonPathConfig,
    result: &mut CLIResult,
) {
    if let Some(ref p) = paths.text {
        if let Some(v) = extract_json_value(data, p) {
            result.text = v.as_str().unwrap_or("").to_string();
        }
    }
    if let Some(ref p) = paths.session_id {
        if let Some(v) = extract_json_value(data, p) {
            result.session_id = v.as_str().map(String::from);
        }
    }
    if let Some(ref p) = paths.model {
        if let Some(v) = extract_json_value(data, p) {
            result.model = v.as_str().unwrap_or("").to_string();
        }
    }
    if let Some(ref p) = paths.input_tokens {
        if let Some(v) = extract_json_value(data, p) {
            result.input_tokens = v.as_u64();
        }
    }
    if let Some(ref p) = paths.output_tokens {
        if let Some(v) = extract_json_value(data, p) {
            result.output_tokens = v.as_u64();
        }
    }
}

pub fn parse_text_output(stdout: &str, stderr: &str, config: &BackendConfig) -> CLIResult {
    let mut result = CLIResult {
        text: stdout.to_string(),
        raw_stdout: stdout.to_string(),
        raw_stderr: stderr.to_string(),
        ..Default::default()
    };

    if let Some(ref pattern) = config.text_patterns.session_id {
        if let Ok(re) = regex::Regex::new(pattern) {
            if let Some(caps) = re.captures(stderr) {
                result.session_id = caps.get(1).map(|m| m.as_str().to_string());
            }
        }
    }

    if let Some(ref pattern) = config.text_patterns.model {
        if let Ok(re) = regex::Regex::new(pattern) {
            if let Some(caps) = re.captures(stdout) {
                result.model = caps.get(1).map(|m| m.as_str().to_string()).unwrap_or_default();
            }
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::TextPatternConfig;

    fn claude_config() -> BackendConfig {
        serde_json::from_str(r#"{
            "command": "claude",
            "args": ["-p"],
            "json_paths": {
                "text": "result",
                "session_id": "session_id",
                "model": "model",
                "input_tokens": "usage.input_tokens",
                "output_tokens": "usage.output_tokens"
            }
        }"#).unwrap()
    }

    #[test]
    fn extract_simple_path() {
        let data = serde_json::json!({"result": "hello", "session_id": "abc"});
        assert_eq!(extract_json_value(&data, "result").unwrap().as_str(), Some("hello"));
    }

    #[test]
    fn extract_nested_path() {
        let data = serde_json::json!({"usage": {"input_tokens": 100}});
        assert_eq!(extract_json_value(&data, "usage.input_tokens").unwrap().as_u64(), Some(100));
    }

    #[test]
    fn extract_missing_returns_none() {
        let data = serde_json::json!({"result": "text"});
        assert!(extract_json_value(&data, "nonexistent.path").is_none());
    }

    #[test]
    fn parse_json_claude_format() {
        let stdout = r#"{"result": "planned", "session_id": "s1", "model": "claude-sonnet-4", "usage": {"input_tokens": 100, "output_tokens": 50}}"#;
        let result = parse_json_output(stdout, &claude_config());
        assert_eq!(result.text, "planned");
        assert_eq!(result.session_id, Some("s1".to_string()));
        assert_eq!(result.model, "claude-sonnet-4");
        assert_eq!(result.input_tokens, Some(100));
        assert_eq!(result.output_tokens, Some(50));
        assert!(!result.parse_error);
    }

    #[test]
    fn parse_json_invalid_falls_back() {
        let result = parse_json_output("not json", &claude_config());
        assert_eq!(result.text, "not json");
        assert!(result.parse_error);
    }

    #[test]
    fn parse_text_with_regex() {
        let config = BackendConfig {
            command: "openclaw".into(),
            output_format: "text".into(),
            text_patterns: TextPatternConfig {
                session_id: Some(r"session[:\s]+([a-f0-9-]+)".into()),
                model: None,
            },
            ..Default::default()
        };
        let result = parse_text_output("done", "session: abc-123 started", &config);
        assert_eq!(result.text, "done");
        assert_eq!(result.session_id, Some("abc-123".to_string()));
    }
}
